### PR TITLE
Creating Minion for Boss Mode

### PR DESCRIPTION
**Pull request by kari0003**
Created basic skeleton for minion
There is no way (ability or function) to create a minion yet.
Feel free to leave this PR open, so i can continue later on ( i will only be back on sunday :S)

### CODE REVIEW

File: wurst/boss/Minion.wurst [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		super(existing, COLL_SIZE_MINION, COLL_SIZE_MINION + 16)
+        selectTarget()
+
+    override function update()
+        super.update()
+        if target.alive != true or getRegionData(target.pos) != currentRegion
+            selectTarget()
+        else if target != null
+            // Spamming the order like this might not be a good idea
+            IssuePointOrder(actor, "move", target.pos.x, target.pos.y)
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: use extension functions

File: wurst/boss/Boss.wurst [hunk 2]
```diff
@@ -78,6 +80,11 @@ public class Boss extends SetupObject
 				for target in tgts
 					new BossMissile(actor,target)
 				destroy tgts
+			case SUMMON_MINION_ID
+				let a = true
+				// TODO `for spawner in minionspawners`
+				new Minion(pos.toVec2(), owner, actor.getFacingAngle())
+				// TODO play sound effect
```
Comment (Frotty) [inline wurst/boss/Boss.wurst]: do todos :P

File: wurst/boss/Minion.wurst [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    construct(vec2 pos, player owner, angle facing)
+        super(pos.toVec3(), owner, getRegionData(pos).getTypeSet().minionId, facing, COLL_SIZE_MINION, COLL_SIZE_MINION + 16)
+        selectTarget()
+
+    construct(vec2 pos, player owner)
+        super(pos.toVec3(), owner, getRegionData(pos).getTypeSet().minionId, COLL_SIZE_MINION, COLL_SIZE_MINION + 16)
+        selectTarget()
+
+    construct(unit existing)
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: this is only used for startmazes, but i guess this could be a general enemy also?

File: wurst/boss/Boss.wurst [hunk 4]
```diff
@@ -16,6 +17,7 @@ constant REAL_DMG_SMALL_ID = 'A001'
 constant AOE_SPELL_ID = 'A06V'
 constant MISSILE_SPELL_ID = 'A06W'
 constant REVIVE_ID = 'A07L'
+constant SUMMON_MINION_ID = 'A085'
 
 public class Boss extends SetupObject
```
Comment (Frotty) [inline wurst/boss/Boss.wurst]: add ability to boss in code so it's independent from map

File: wurst/boss/Minion.wurst [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        super(existing, COLL_SIZE_MINION, COLL_SIZE_MINION + 16)
+        selectTarget()
+
+    override function update()
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: I'm not sure if ordering each tick is best performance wise. Maybe just order an attack?

File: wurst/boss/Minion.wurst [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        super.update()
+        if target.alive != true or getRegionData(target.pos) != currentRegion
+            selectTarget()
+        else if target != null
+            // Spamming the order like this might not be a good idea
+            IssuePointOrder(actor, "move", target.pos.x, target.pos.y)
+
+    function selectTarget()
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: add secret zombie cheese item that throws bait to lure them away :3

File: wurst/boss/Minion.wurst (hunk 3 above)
Comment (kari0003) [inline wurst/boss/Minion.wurst]: yes indeed it could be, however in any case; we dont need this constructor.

File: wurst/boss/Boss.wurst (hunk 4 above)
Comment (kari0003) [inline wurst/boss/Boss.wurst]: Where should i add it? .j file?

Comment (Frotty) [inline wurst/boss/Boss.wurst]: in the Boss constructor

File: wurst/boss/Minion.wurst [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import CustomBar
+import ClosureForGroups
+import Fx
+
+public constant COLL_SIZE_MINION = 0.
+public constant ATTACK_DISTANCE = 70.
+public constant ATTACK_RANGE = 120.
+public constant ATTACK_DAMAGE = 100.
+
+class MinionData
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: Class not necessary if it only has one member

File: wurst/boss/Minion.wurst [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        hpBar.setForeground(colorA(200,255,125,255))
+        
+    function onDamage(Escaper escaper, real dmg)
+        currentHP -= dmg
+        if currentHP < 1
+            actor.kill()
+            setup.abortOrder()
+            hpBar.fadeOut(1.5, true, 0, 0)
+            sleeps = true
+            getTimer()..start(3, function Minion.removeAfterDeath)..setData(new MinionData(this) castTo int)
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: Use doAfter() for the timed stuff

File: wurst/boss/Minion.wurst [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+    override function update()
+        super.update()
+        hpBar.setPosXY(pos.x - 50, pos.y + 50)
+        hpBar.setValue(currentHP / maxHp * 100.0)
+        if target == null or target.alive != true or target.getCurrentRegion() != this.getCurrentRegion()
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: target.alive != true -> not target.alive

File: wurst/boss/Minion.wurst [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        if currentHP < 1
+            actor.kill()
+            setup.abortOrder()
+            hpBar.fadeOut(1.5, true, 0, 0)
+            sleeps = true
+            getTimer()..start(3, function Minion.removeAfterDeath)..setData(new MinionData(this) castTo int)
+
+    override function update()
+        super.update()
+        hpBar.setPosXY(pos.x - 50, pos.y + 50)
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: use casecade operator (..)

File: wurst/boss/Minion.wurst [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        hpBar.setValue(currentHP / maxHp * 100.0)
+        if target == null or target.alive != true or target.getCurrentRegion() != this.getCurrentRegion()
+            selectTarget()
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: I think it is not a good idea to check for targets every tick. 
You could instead increase collisionsize and wake the minion up using onSensor (from the class Minion extends)
Or you add a buff to the minion that checks for targets - buffs are only updated 4 times per second.

File: wurst/boss/Minion.wurst [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        else if target != null
+            if attacking
+                actor.setFacing(pos.angleTo2d(target.pos))
+            if not attacking and target.pos.distanceTo(pos) < ATTACK_DISTANCE
+                attack()
+            else if not attacking
+                // Spamming the order like this might not be a good idea
+                setup.issuePointOrder("move", target.pos.toVec2())
+
+    function selectTarget()
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: It would be nice to have a maximum distance to search for targets

File: wurst/boss/Minion.wurst [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+    override function update()
+        super.update()
+        hpBar.setPosXY(pos.x - 50, pos.y + 50)
+        hpBar.setValue(currentHP / maxHp * 100.0)
+        if target == null or target.alive != true or target.getCurrentRegion() != this.getCurrentRegion()
+            selectTarget()
+        else if target != null
+            if attacking
+                actor.setFacing(pos.angleTo2d(target.pos))
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: Please add a maximum guard distance or some other way to stop the minion chasing the escapers. Maybe there should be the posibilitiy of a safe zone somehow

File: wurst/boss/Minion.wurst [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+        var t = GetExpiredTimer()
+        MinionData data = t.getData() castTo MinionData
+        let minion = data.entity castTo Minion
+        if minion.currentHP > 0
+            let facing = data.entity.actor.getFacingAngle()
+            let tpos = minion.getPos().toVec2().polarOffset(facing, 80)
+            destroy new Fx(tpos, facing, Other.manaflareboltimpact)..setZ(30)..setScale(1.1)
+            minion.attacking = false
+            forUnitsInRange(tpos, ATTACK_RANGE, (unit u) -> begin
+                let entity = u.getEntity()
```
Comment (Frotty) [inline wurst/boss/Minion.wurst]: plz add kewl spec1al effx and texttags and shit 👍 👍

File: wurst/boss/WeaponTypes.wurst [hunk 15]
```diff
@@ -14,15 +15,23 @@ public interface WeaponListener
 
 public WeaponType array weapontypes
 
+function damageTarget(Entity entity, Escaper escaper, real damage)
```
Comment (Frotty) [inline wurst/boss/WeaponTypes.wurst:18]: not good if we have to extend this one function for everything that should be procced by damage.
We should do this in another way, maybe via buffs or listeners

**Review by Frotty — CHANGES_REQUESTED** [review]

**Review by Frotty — APPROVED** [review]
Thanks krarolinorio

**Pull request merged**